core/src/query.rs

```diff
@@ -75,3 +124,149 @@ pub struct MongoColMetadata {
+    pub additional_properties: Option<bool>,
+}
+
+impl SqlGetResultSchemaResponse {
+    /// Converts a sqlGetResultSchema command response into a list of column
+    /// metadata. Ensures the top-level schema is an Object with properties,
+    /// and ensures the same for each top-level property -- which correspond
+    /// to datasources. The metadata is sorted alphabetically by datasource
+    /// name and then by field name. As in, a result set with schema:
+    ///
+    ///   {
+    ///     bsonType: "object",
+    ///     properties: {
+    ///       "foo": {
+    ///         bsonType: "object",
+    ///         properties: { "b": { bsonType: "int" }, "a": { bsonType: "string" } }
+    ///       },
+    ///       "bar": {
+    ///         bsonType: "object",
+    ///         properties: { "c": { bsonType: "int" } }
+    ///       }
+    ///   }
+    ///
+    /// produces a list of metadata with the order: "bar.c", "foo.a", "foo.b".
+    fn process_metadata(self) -> Result<Vec<MongoColMetadata>> {
+        let result_set_schema: SimplifiedJsonSchema = self.schema.json_schema.into();
+        result_set_schema.assert_datasource_schema()?;
+
+        result_set_schema
+            // 1. Access result_set_schema.properties and sort alphabetically.
+            //    This means we are sorting by datasource name.
+            .properties
+            .unwrap()
+            .into_iter()
+            .sorted_by(|a, b| Ord::cmp(&a.0, &b.0))
+            // 2. Flat-map fields for each datasource, sorting fields alphabetically.
+            .flat_map(|(datasource_name, datasource_schema)| {
+                datasource_schema
+                    .properties
+                    .unwrap()
+                    .into_iter()
+                    .sorted_by(|a, b| Ord::cmp(&a.0, &b.0))
+                    .map(|(field_name, field_schema)| {
+                        (datasource_name, datasource_schema, field_name, field_schema)
+                    })
+            })
+            // 3. Map each field into a MongoColMetadata.
+            .map(
+                |(datasource_name, datasource_schema, field_name, field_schema)| {
+                    datasource_schema.assert_datasource_schema()?;
+
+                    let field_nullability =
+                        datasource_schema.get_field_nullability(field_name.clone())?;
+
+                    Ok(Self::create_column_metadata(
+                        datasource_name,
+                        field_name,
+                        *field_schema,
+                        field_nullability,
+                    ))
+                },
+            )
+            // 4. Collect as a Vec.
+            .collect::<Result<Vec<MongoColMetadata>>>()
+    }
+
+    fn create_column_metadata(
+        datasource_name: String,
+        field_name: String,
+        field_schema: SimplifiedJsonSchema,
+        is_nullable: bool,
+    ) -> MongoColMetadata {
+        todo!()
```

Comment:
This and the following 2 `todo!()`s are the follow-up work I'm doing right now as you review the rest of this PR!